### DISCUSSION TITLE
Does the `new` statement introduce a scheduling point?

### DISCUSSION THREAD

**Discussion by dotnwat** [p-org/P#765] [Q&A] [answered]
In the documentation for the `new` statement (https://p-org.github.io/P/manual/statements/#new) the language used:

> delivered as a payload to the entry function of the start state of the created machine

makes it sound as though a message is delivered to the fifo queue of the new machine. Is this true? Is a scheduling point introduced, or does the entry function of the new machine run inline with the caller of `new`?

**Comment by AnDongLi** [comment, marked as answer]
For PChecker, new statement compiled to one of CreateStateMachine methods in ControlledRuntime, there is a scheduled point just before creating a StateMachine: 
[https://github.com/p-org/P/blob/b45194b2fa9e7334bf488e3ac887227ee1c2c2d3/Src/PChecker/CheckerCore/SystematicTesting/ControlledRuntime.cs#L498](Code url)

Right after the StateMachine is created, OnEntry action will be called during the first time the newly created StateMachine instance runs its EventHandler:
[https://github.com/p-org/P/blob/b45194b2fa9e7334bf488e3ac887227ee1c2c2d3/Src/PChecker/CheckerCore/SystematicTesting/ControlledRuntime.cs#L685](Code url)

In generated ClientServer.cs, Init State of a StateMachine looks like this and Anon is an anonymous function generated for action onEntry and it will be called during above mentioned initialization.

        [Start]
        [OnEntry(nameof(Anon))]
        class Init : State
        {
        }

**Comment by ankushdesai** [comment]
A scheduling point is introduced between the machine being created and it executing the entry function of the start state
